pkg/gardenlet/controller/extensions/extensions_sync_control.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	gardenextensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	gardenextensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/extensions/listers/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ExtensionStateSyncController replicates the extensions' resources states in Shoot's ShootState resource
+type ExtensionStateSyncController struct {
+	k8sGardenClient      kubernetes.Interface
+	seedClient           kubernetes.Interface
+	extensionsInformers  gardenextensionsinformers.SharedInformerFactory
+	extensionsSeedClient gardenextensionsclient.Interface
+
+	context             context.Context
+	cancelFunc          context.CancelFunc
+	log                 *logrus.Entry
+	infrastructuresList v1alpha1.InfrastructureLister
+	workersList         v1alpha1.WorkerLister
+
+	recorder record.EventRecorder
+
+	infrastructuresHasSynced cache.InformerSynced
+	workerHasSynced          cache.InformerSynced
```

Comment:
Any reason that we only have these two? What about the others? Or is it because it's WIP?

Reply:
It's because WIP. I've suggested to move this + the worker logic in a separate struct/structures  if the controller creation seems like too big of a method.

Reply:
Yeah, but also the GEP states that most of the extension resources are irrelevant for sync, except the Extensions resource - this I am not sure how we would like to replicate in the shootstate.